# Checkpoint Two: Exploratory Data Analysis

Now that your chosen dataset is approved, it is time to start working on your analysis. Use this notebook to perform your EDA and make notes where directed to as you work.

## Getting Started

Since we have not provided your dataset for you, you will need to load the necessary files in this repository. Make sure to include a link back to the original dataset here as well.

My dataset:

Your first task in EDA is to import necessary libraries and create a dataframe(s). Make note in the form of code comments of what your thought process is as you work on this setup task.

<h1 style="background-color:SlateBlue; color:White; text-align:center">Resources</h1>

- [20 Must-Know Pandas Function for Exploratory Data Analysis](https://www.analyticsvidhya.com/blog/2021/04/20-must-known-pandas-function-for-exploratory-data-analysis-eda/)
- [A Comprehensive Guide to Data Analysis using Pandas](https://www.analyticsvidhya.com/blog/2021/05/a-comprehensive-guide-to-data-analysis-using-pandas-hands-on-data-analysis-on-imdb-movies-data/)
- [Python | Pandas dataframe.groupby()](https://www.geeksforgeeks.org/python-pandas-dataframe-groupby/)

In [52]:
import pandas as pd

items = pd.read_csv("roll_20_items.csv")
monsters = pd.read_csv("cleaned_monsters_basic.csv")
more_monsters = pd.read_csv("aidedd_blocks2.csv")

# I have 3 datasets to explore.  I want to look at each one at a time.

## Get to Know the Numbers

Now that you have everything setup, put any code that you use to get to know the dataframe and its rows and columns better in the cell below. You can use whatever techniques you like, except for visualizations. You will put those in a separate section.

When working on your code, make sure to leave comments so that your mentors can understand your thought process.

<h1 style="background-color:DodgerBlue; color:White; text-align:center">Items Data</h1>

In [40]:
items.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 771 entries, 0 to 770
Data columns (total 17 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   item              771 non-null    object 
 1   type              771 non-null    object 
 2   notes             338 non-null    object 
 3   item_rarity       382 non-null    object 
 4   weight            610 non-null    float64
 5   attunement        771 non-null    int64  
 6   details           702 non-null    object 
 7   modifiers         186 non-null    object 
 8   ac                194 non-null    float64
 9   save              5 non-null      object 
 10  stealth           112 non-null    object 
 11  damage            204 non-null    object 
 12  damage_type       201 non-null    object 
 13  duration          11 non-null     object 
 14  secondary damage  31 non-null     object 
 15  range             88 non-null     object 
 16  properties        179 non-null    object 
dt

In [41]:
items.describe()

weight  attunement          ac
count  610.000000  771.000000  194.000000
mean    12.898496    0.337224   13.443299
std     18.112379    0.473069    3.301653
min      0.000000    0.000000    2.000000
25%      2.000000    0.000000   12.000000
50%      4.000000    0.000000   14.000000
75%     13.000000    1.000000   15.000000
max     70.000000    1.000000   18.000000

Only 3 coluumns have numbers.  

- **Weight**: 610/771 items have significant weight with an average of ~13 pounds
- **Attunement**: every item has a 1 (attunement) or 0 (no attunement), with an average of ~30% requiring attunement
- **AC**: 194/771 items improve AC, with an average of ~13 points

In [42]:
items.head()

item              type                          notes  \
0                abacus  adventuring gear                            NaN   
1                  acid  adventuring gear                     consumable   
2      adamantine armor             armor  medium or heavy, but not hide   
3      alchemist's fire  adventuring gear                     consumable   
4  alchemist's supplies  adventuring gear                artisan's tools   

  item_rarity  weight  attunement  \
0         NaN     2.0           0   
1         NaN     1.0           0   
2         NaN     NaN           0   
3         NaN     1.0           0   
4    standard     8.0           0   

                                             details modifiers  ac       save  \
0                                                NaN       NaN NaN        NaN   
1  As an action, you can splash the contents of t...       NaN NaN        NaN   
2  This suit of armor is reinforced with adamanti...       NaN NaN        NaN   
3  This sticky, adhesive fluid ignites when expos...       NaN NaN  dexterity   
4  These Special tools include the items needed t...       NaN NaN        NaN   

  stealth damage damage_type duration secondary damage  range  \
0     NaN    NaN         NaN      NaN              NaN    NaN   
1     NaN    2d6        acid      NaN              NaN  20/60   
2     NaN    NaN         NaN      NaN              NaN    NaN   
3     NaN    1d4        fire      NaN              NaN  20/60   
4     NaN    NaN         NaN      NaN              NaN    NaN   

           properties  
0                 NaN  
1  improvised weapons  
2                 NaN  
3  improvised weapons  
4                 NaN

In [43]:
items.columns

Index(['item', 'type', 'notes', 'item_rarity', 'weight', 'attunement',
       'details', 'modifiers', 'ac', 'save', 'stealth', 'damage',
       'damage_type', 'duration', 'secondary damage', 'range', 'properties'],
      dtype='object')

Looking at **head** and **columns**, I can see how things are organized.

- **Item** (string) = an object's name
- **Type** (string) = its purpose in the game
- **Notes** (string) = attributes or how it is used
- **Item rarity** (string) = reflects the scarity of an object as well as its power.  The more rare (aka legendary, then the more powerful, but also smaller amount produced)  
    - Standard (see alchemist tools) come with that class.  
    - common --> uncommon --> rare --> very rare --> legendary and artefact.
- **weight** (float) = how heavy something is.  This is useful when adding up all the items, money, and weapons your character is holding.  Each race has a range of weight they can easily manage.  There are tiers for being overweight that change your movement, depending on how large the discrepancy is between allowed weight and actual weight.  
- **attunement** (int) = magicial items especially need to "attune" to the user in order to be used.  
- **details** (string) = requirements and extra info such as bonuses, perks, how to use notes
- **modifiers** (string) = any changes in the players current stats.  Could be positive or negative.
- **ac** (float) = stands for "Armor Class" reflects the total amount of base protection a character has from damage.  Typically accumulated from armor, shields, or magical items.
- **save** (string) = some items in the game require the character to check and see if they beat it or not.  Typically, a save is a die/dice roll compared to a specific stat.  If the character meets or beats the dice roll, then they either avoid the harm or take on the good thing (and vise versa if they miss the roll).
- **stealth** (string) = how sneaky something is.  Leather armor is a lot less noisy than full plate, which can cause a disadvantage in a sneaky situation.  
- **damage** (string) = The total possible amount of damage a weapon or spell can cause. a die value (sometimes die value + a skill modifier).
- **damage_type** (string) = there are different types of damage that is related to a weapon, or if magical an element.  A spear causes piercing damage, a club bludgeoning, and a fireball fire damage.  Certain gear, items, races, classes, monsters are immume to some damage types and others more vulenrable.
- **duration** (string) = how long an effect lasts.  if a spell has been successfully cast, it may last hours, days or seconds. 
- **secondary damage** (string) = is seen with two-handed weapons.  if using with 2 hands, you are stronger therefore causing more damgae.  
- **range** (string) = how close or far something is for a ranged character to successfully hit its target.  Some ranged weapons have a normal range and long range, when using long range you are at disadvantage
- **properties** (string) = weapon properties are how the weapon is used, and any extra steps it needs.  For example, a blowgun has range and loading properties.  meaning that after you fire (range) your will need to reload it (loading).

In [44]:
items.isnull().sum()

item                  0
type                  0
notes               433
item_rarity         389
weight              161
attunement            0
details              69
modifiers           585
ac                  577
save                766
stealth             659
damage              567
damage_type         570
duration            760
secondary damage    740
range               683
properties          592
dtype: int64

There are a lot of nulls (NaN) values in the items dataframe.  

Not sure how to best handle these yet.

In [45]:
items.groupby('item')[['weight']].head()

weight
0       2.0
1       1.0
2       NaN
3       1.0
4       8.0
..      ...
766     NaN
767     5.0
768     6.0
769     4.0
770     1.0

[771 rows x 1 columns]

In [46]:
items.groupby("secondary damage")[["weight"]].count()

weight
secondary damage        
1d10                  14
1d8                   13
2d10                   1
2d6                    3

In [47]:
damage= items.groupby("damage")
damage.first()

#https://www.geeksforgeeks.org/python-pandas-dataframe-groupby/ <-- playing with this

item              type              notes item_rarity  \
damage                                                                      
0                 net +1     ranged weapon               None        None   
1                blowgun     ranged weapon   martial, blowgun    standard   
1d10              glaive      melee weapon    martial, glaive    standard   
1d12            greataxe      melee weapon       martial, axe    standard   
1d4     alchemist's fire  adventuring gear         consumable    standard   
1d6        hand crossbow     ranged weapon  martial, crossbow    standard   
1d8            battleaxe      melee weapon       martial, axe    standard   
2d6                 acid  adventuring gear         consumable    standard   
5                    oil  adventuring gear               None    standard   

        weight  attunement                                            details  \
damage                                                                          
0          3.0           0  You have a +1 bonus to Attack Rolls made with ...   
1          1.0           0  Ammunition: You can use a weapon that has the ...   
1d10       6.0           0  You have a +1 bonus to Attack and Damage Rolls...   
1d12       7.0           0  You have a +1 bonus to Attack and Damage Rolls...   
1d4        1.0           0  This sticky, adhesive fluid ignites when expos...   
1d6        3.0           0  Ammunition: You can use a weapon that has the ...   
1d8        4.0           0  You have a +1 bonus to Attack and Damage Rolls...   
2d6        1.0           0  As an action, you can splash the contents of t...   
5          1.0           0  Oil usually comes in a clay flask that holds 1...   

                                  modifiers  ac       save stealth  \
damage                                                               
0       ranged attacks +1, ranged damage +1 NaN       None    None   
1       ranged attacks +1, ranged damage +1 NaN       None    None   
1d10      melee attacks +1, melee damage +1 NaN       None    None   
1d12      melee attacks +1, melee damage +1 NaN       None    None   
1d4       melee attacks +1, melee damage +1 NaN  dexterity    None   
1d6     ranged attacks +1, ranged damage +1 NaN       None    None   
1d8       melee attacks +1, melee damage +1 NaN       None    None   
2d6       melee attacks +1, melee damage +1 NaN       None    None   
5                                      None NaN       None    None   

       damage_type  duration secondary damage    range  \
damage                                                   
0             None      None             None     5/15   
1         piercing    1 hour             None   25/100   
1d10      slashing      None             None  100/400   
1d12      slashing      None             None     None   
1d4           fire  1 minute             2d10    20/60   
1d6       piercing      None              2d6   30/120   
1d8       slashing      None             1d10    20/60   
2d6           acid      None              2d6    20/60   
5             fire  2 rounds             None    20/60   

                      properties  
damage                            
0                  range, thrown  
1                 loading, range  
1d10    heavy, reach, two-handed  
1d12           heavy, two-handed  
1d4           improvised weapons  
1d6        light, loading, range  
1d8                    versatile  
2d6           improvised weapons  
5             improvised weapons

In [71]:
items.nlargest(5, "weight")

item              type               notes  \
27                     barrel  adventuring gear                 NaN   
12   armor of invulnerability       heavy armor               armor   
156               demon armor       heavy armor  cursed item, armor   
169             dwarven plate       heavy armor               armor   
426               plate armor       heavy armor               armor   

    item_rarity  weight  attunement  \
27          NaN    70.0           0   
12    legendary    65.0           1   
156   very rare    65.0           1   
169   very rare    65.0           0   
426    standard    65.0           0   

                                               details modifiers    ac save  \
27   A barrel can hold 40 gallons of liquid or 4 cu...       NaN   NaN  NaN   
12   Requires Attunement\n\nYou have Resistance to ...       NaN  18.0  NaN   
156  Requires Attunement\n\nWhile wearing this armo...     ac +1  18.0  NaN   
169  While wearing this armor, you gain a +2 bonus ...     ac +2  18.0  NaN   
426  Plate consists of shaped, interlocking metal p...       NaN  18.0  NaN   

          stealth damage damage_type duration secondary damage range  \
27            NaN    NaN         NaN      NaN              NaN   NaN   
12   disadvantage    NaN         NaN      NaN              NaN   NaN   
156  disadvantage    NaN         NaN      NaN              NaN   NaN   
169  disadvantage    NaN         NaN      NaN              NaN   NaN   
426  disadvantage    NaN         NaN      NaN              NaN   NaN   

    properties  
27         NaN  
12         NaN  
156        NaN  
169        NaN  
426        NaN

In [72]:
items.nunique()

item                771
type                 16
notes                76
item_rarity           9
weight               32
attunement            2
details             379
modifiers            25
ac                    9
save                  2
stealth               1
damage                9
damage_type           7
duration              7
secondary damage      4
range                12
properties           27
dtype: int64

### Ideas to Explore in the Items df
- weapons: damage, magical/nonmagical, type, range/non-range
- magic items: damage? buffs? attunements?

<h1 style="background-color:Tomato; color:White; text-align:center">Monsters Data</h1>

In [62]:
monsters.head()

Unnamed: 0                name    size         monster_type      alignment  \
0           0             Aboleth   Large           Aberration    lawful evil   
1           1             Acolyte  Medium  Humanoid (any race)  any alignment   
2           2  Adult Black Dragon    Huge               Dragon   chaotic evil   
3           3   Adult Blue Dragon    Huge               Dragon    lawful evil   
4           4  Adult Brass Dragon    Huge               Dragon   chaotic good   

   ac   hp  strength  str_mod  dex  ...  perception  stealth  persuasion  \
0  17  135        21        5    9  ...          10        0           0   
1  10    9        10        0   10  ...           0        0           0   
2  19  195        23        6   14  ...          11        7           0   
3  19  225        25        7   10  ...          12        5           0   
4  18  172        23        6   10  ...          11        5           8   

   insight  deception  arcana  religion  acrobatics  athletics intimidation  
0        0          0       0         0           0          0            0  
1        0          0       0         0           0          0            0  
2        0          0       0         0           0          0            0  
3        0          0       0         0           0          0            0  
4        0          0       0         0           0          0            0  

[5 rows x 45 columns]

In [54]:
monsters.shape

(324, 45)

In [57]:
monsters.describe()

Unnamed: 0          ac          hp    strength     str_mod         dex  \
count   324.00000  324.000000  324.000000  324.000000  324.000000  324.000000   
mean    161.50000   14.043210   81.324074   15.117284    2.169753   12.651235   
std      93.67497    3.245094   99.182494    6.739180    3.227045    3.210658   
min       0.00000    5.000000    1.000000    1.000000   -5.000000    1.000000   
25%      80.75000   12.000000   18.750000   11.000000    0.000000   10.000000   
50%     161.50000   13.000000   45.000000   16.000000    3.000000   13.000000   
75%     242.25000   17.000000  114.000000   19.000000    4.000000   15.000000   
max     323.00000   25.000000  676.000000   30.000000    9.000000   28.000000   

          dex_mod         con     con_mod       intel  ...  perception  \
count  324.000000  324.000000  324.000000  324.000000  ...  324.000000   
mean     1.123457   15.067901    2.283951    7.799383  ...    1.601852   
std      1.605970    4.495097    2.183442    5.715104  ...    3.913001   
min     -5.000000    8.000000   -1.000000    1.000000  ...    0.000000   
25%      0.000000   12.000000    1.000000    2.000000  ...    0.000000   
50%      1.000000   14.000000    2.000000    7.000000  ...    0.000000   
75%      2.000000   18.000000    3.250000   12.000000  ...    0.000000   
max      9.000000   30.000000    9.000000   25.000000  ...   17.000000   

          stealth  persuasion     insight   deception      arcana    religion  \
count  324.000000  324.000000  324.000000  324.000000  324.000000  324.000000   
mean     0.685185    0.246914    0.299383    0.231481    0.259259    0.061728   
std      1.920952    1.631963    1.519695    1.323331    1.667595    0.877594   
min      0.000000    0.000000    0.000000    0.000000    0.000000    0.000000   
25%      0.000000    0.000000    0.000000    0.000000    0.000000    0.000000   
50%      0.000000    0.000000    0.000000    0.000000    0.000000    0.000000   
75%      0.000000    0.000000    0.000000    0.000000    0.000000    0.000000   
max      9.000000   16.000000   10.000000   11.000000   18.000000   15.000000   

       acrobatics   athletics  intimidation  
count  324.000000  324.000000    324.000000  
mean     0.018519    0.185185      0.015432  
std      0.333333    1.415186      0.277778  
min      0.000000    0.000000      0.000000  
25%      0.000000    0.000000      0.000000  
50%      0.000000    0.000000      0.000000  
75%      0.000000    0.000000      0.000000  
max      6.000000   14.000000      5.000000  

[8 rows x 38 columns]

In [58]:
monsters.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 324 entries, 0 to 323
Data columns (total 45 columns):
 #   Column                    Non-Null Count  Dtype 
---  ------                    --------------  ----- 
 0   Unnamed: 0                324 non-null    int64 
 1   name                      324 non-null    object
 2   size                      324 non-null    object
 3   monster_type              324 non-null    object
 4   alignment                 324 non-null    object
 5   ac                        324 non-null    int64 
 6   hp                        324 non-null    int64 
 7   strength                  324 non-null    int64 
 8   str_mod                   324 non-null    int64 
 9   dex                       324 non-null    int64 
 10  dex_mod                   324 non-null    int64 
 11  con                       324 non-null    int64 
 12  con_mod                   324 non-null    int64 
 13  intel                     324 non-null    int64 
 14  int_mod                   

<h1 style="background-color:MediumSeaGreen; color:White; text-align:center">More Monster Data</h1>

this has MORE monster data from more sources

In [55]:
more_monsters.shape

(762, 53)

In [59]:
more_monsters.describe()

ac          hp          cr       speed        swim         fly  \
count  762.000000  762.000000  762.000000  762.000000  762.000000  762.000000   
mean    14.595801   88.194226    5.546424   29.803150    4.960630   13.661417   
std      3.095821   94.779066    6.252214   12.334075   14.447732   27.694709   
min      5.000000    1.000000    0.000000    0.000000    0.000000    0.000000   
25%     12.000000   22.000000    0.500000   30.000000    0.000000    0.000000   
50%     14.000000   58.000000    3.000000   30.000000    0.000000    0.000000   
75%     17.000000  126.000000    9.000000   40.000000    0.000000    0.000000   
max     25.000000  676.000000   30.000000  120.000000  120.000000  150.000000   

            climb      burrow  passive_perception  darkvision  ...  \
count  762.000000  762.000000          762.000000  762.000000  ...   
mean     2.952756    0.984252           13.023622   47.755906  ...   
std      9.217499    5.322924            3.908879   46.165198  ...   
min      0.000000    0.000000            0.000000    0.000000  ...   
25%      0.000000    0.000000           10.000000    0.000000  ...   
50%      0.000000    0.000000           12.000000   60.000000  ...   
75%      0.000000    0.000000           14.000000   60.000000  ...   
max     40.000000   50.000000           36.000000  240.000000  ...   

          stealth  persuasion     insight   deception      arcana    religion  \
count  762.000000  762.000000  762.000000  762.000000  762.000000  762.000000   
mean     1.639108    0.385827    0.520997    0.577428    0.750656    0.305774   
std      2.703906    1.861346    2.026164    2.134730    2.625215    1.518936   
min      0.000000    0.000000    0.000000    0.000000    0.000000    0.000000   
25%      0.000000    0.000000    0.000000    0.000000    0.000000    0.000000   
50%      0.000000    0.000000    0.000000    0.000000    0.000000    0.000000   
75%      4.000000    0.000000    0.000000    0.000000    0.000000    0.000000   
max     16.000000   16.000000   14.000000   15.000000   22.000000   17.000000   

       acrobatics   athletics  intimidation   legendary  
count  762.000000  762.000000    762.000000  762.000000  
mean     0.144357    0.400262      0.392388    0.056430  
std      0.983518    1.738079      1.898777    0.230903  
min      0.000000    0.000000      0.000000    0.000000  
25%      0.000000    0.000000      0.000000    0.000000  
50%      0.000000    0.000000      0.000000    0.000000  
75%      0.000000    0.000000      0.000000    0.000000  
max     11.000000   14.000000     18.000000    1.000000  

[8 rows x 43 columns]

In [60]:
more_monsters.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 762 entries, 0 to 761
Data columns (total 53 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   name                762 non-null    object 
 1   size                762 non-null    object 
 2   type                762 non-null    object 
 3   alignment           762 non-null    object 
 4   languages           762 non-null    object 
 5   ac                  762 non-null    float64
 6   hp                  762 non-null    int64  
 7   cr                  762 non-null    float64
 8   speed               762 non-null    int64  
 9   swim                762 non-null    int64  
 10  fly                 762 non-null    int64  
 11  climb               762 non-null    int64  
 12  burrow              762 non-null    int64  
 13  passive_perception  762 non-null    int64  
 14  darkvision          762 non-null    int64  
 15  truesight           762 non-null    int64  
 16  tremorse

In [61]:
more_monsters.head()

name    size                  type      alignment  \
0        aarakocra  medium  humanoid (aarakocra)   neutral good   
1          abjurer  medium   humanoid (any race)  any alignment   
2          aboleth   large            aberration    lawful evil   
3  abominable-yeti    huge           monstrosity   chaotic evil   
4         acererak  medium                undead   neutral evil   

                                           languages    ac   hp     cr  speed  \
0                                   Auran, Aarakocra  12.0   13   0.25     20   
1                                 any four languages  12.0   84   9.00     30   
2                     Deep Speech, telepathy 120 ft.  17.0  135  10.00     10   
3                                               Yeti  15.0  137   9.00     40   
4  Abyssal, Common, Draconic, Dwarvish, Elvish, G...  21.0  285  23.00     30   

   swim  ...  religion  acrobatics  athletics  intimidation  \
0     0  ...         0           0          0             0   
1     0  ...         0           0          0             0   
2    40  ...         0           0          0             0   
3     0  ...         0           0          0             0   
4     0  ...        15           0          0             0   

                                      senses  \
0                      passive Perception 15   
1                      passive Perception 11   
2  darkvision 120 ft., passive Perception 20   
3   darkvision 60 ft., passive Perception 15   
4   truesight 120 ft., passive Perception 22   

                                          attributes  \
0  Dive Attack. If the aarakocra is flying and di...   
1                                        unavailable   
2  Amphibious. The aboleth can breathe air and wa...   
3                                        unavailable   
4                                        unavailable   

                                             actions  \
0  Talon. Melee Weapon Attack: +4 to hit, reach 5...   
1                                        unavailable   
2  Multiattack. The aboleth makes three tentacle ...   
3                                        unavailable   
4                                        unavailable   

                                   legendary_actions  legendary  \
0                                               none          0   
1                                        unavailable          0   
2  The aboleth can take 3 legendary actions, choo...          1   
3                                        unavailable          0   
4                                        unavailable          0   

                              source  
0                Monster Manual (BR)  
1           Volo's Guide to Monsters  
2               Monster Manual (SRD)  
3                     Monster Manual  
4  Adventures (Tomb of Annihilation)  

[5 rows x 53 columns]

In [63]:
more_monsters.isnull().sum()

name                   0
size                   0
type                   0
alignment              0
languages              0
ac                     0
hp                     0
cr                     0
speed                  0
swim                   0
fly                    0
climb                  0
burrow                 0
passive_perception     0
darkvision             0
truesight              0
tremorsense            0
blindsight             0
strength               0
str_mod                0
dex                    0
dex_mod                0
con                    0
con_mod                0
intel                  0
int_mod                0
wis                    0
wis_mod                0
cha                    0
cha_mod                0
str_save               0
dex_save               0
con_save               0
int_save               0
wis_save               0
cha_save               0
history                0
perception             0
stealth                0
persuasion             0


In [65]:
more_monsters.nlargest(5, "cr")

name        size                 type     alignment  \
643   tarrasque  gargantuan  monstrosity (titan)     unaligned   
652      tiamat  gargantuan                fiend  chaotic evil   
169  demogorgon        huge        fiend (demon)  chaotic evil   
502       orcus        huge        fiend (demon)  chaotic evil   
759      zariel       large        fiend (devil)   lawful evil   

                      languages    ac   hp    cr  speed  swim  ...  religion  \
643                        none  25.0  676  30.0     40     0  ...         0   
652  Common, Draconic, Infernal  25.0  615  30.0     60     0  ...        17   
169      all, telepathy 120 ft.  22.0  406  26.0     50    50  ...         0   
502      all, telepathy 120 ft.  17.0  405  26.0     40     0  ...         0   
759      all, telepathy 120 ft.  21.0  580  26.0     50     0  ...         0   

     acrobatics  athletics  intimidation  \
643           0          0             0   
652           0          0             0   
169           0          0             0   
502           0          0             0   
759           0          0            18   

                                                senses  \
643          blindsight 120 ft., passive Perception 10   
652  darkvision 240 ft., truesight 120 ft., passive...   
169           truesight 120 ft., passive Perception 29   
502           truesight 120 ft., passive Perception 22   
759          darkvision 120 ft., passive Perception 26   

                                            attributes  \
643  Legendary Resistance (3/Day). If the tarrasque...   
652                                        unavailable   
169                                        unavailable   
502  Wand of Orcus. The wand has 7 charges, and any...   
759  Devil's Sight. Magical darkness doesn't impede...   

                                               actions  \
643  Multiattack. The tarrasque can use its Frightf...   
652                                        unavailable   
169                                        unavailable   
502  Multiattack. Orcus makes two Wand of Orcus att...   
759  Multiattack. Zariel attacks twice with her lon...   

                                     legendary_actions  legendary  \
643  The tarrasque can take 3 legendary actions, ch...          1   
652                                        unavailable          0   
169                                        unavailable          0   
502  Orcus can take 3 legendary actions, choosing f...          1   
759  Zariel can take 3 legendary actions, choosing ...          1   

                              source  
643             Monster Manual (SRD)  
652  Adventures (Tyranny of Dragons)  
169      Mordenkainen's Tome of Foes  
502      Mordenkainen's Tome of Foes  
759      Mordenkainen's Tome of Foes  

[5 rows x 53 columns]

In [66]:
more_monsters.corr()

ac        hp        cr     speed      swim  \
ac                  1.000000  0.668608  0.707400  0.259286  0.074267   
hp                  0.668608  1.000000  0.926309  0.264551  0.185713   
cr                  0.707400  0.926309  1.000000  0.224442  0.125081   
speed               0.259286  0.264551  0.224442  1.000000 -0.053137   
swim                0.074267  0.185713  0.125081 -0.053137  1.000000   
fly                 0.334696  0.365880  0.354818 -0.015583  0.058982   
climb               0.005731 -0.005170 -0.026808  0.091807 -0.092373   
burrow              0.203993  0.212331  0.158885  0.130051  0.005631   
passive_perception  0.551881  0.638400  0.679880  0.232996  0.132180   
darkvision          0.299636  0.245644  0.249490  0.043879  0.067542   
truesight           0.272475  0.320405  0.409311  0.076138 -0.015778   
tremorsense         0.127235  0.100349  0.054065  0.043802 -0.030995   
blindsight          0.133651  0.245539  0.193788 -0.022133  0.152678   
strength            0.546298  0.707779  0.630017  0.422258  0.199649   
str_mod             0.480799  0.581863  0.535957  0.403241  0.132042   
dex                 0.275833  0.047315  0.144512  0.119597  0.007816   
dex_mod             0.284240  0.055722  0.154142  0.126295  0.016815   
con                 0.625748  0.842333  0.791179  0.315804  0.181854   
con_mod             0.579543  0.768436  0.735387  0.285986  0.105523   
intel               0.485368  0.451116  0.562895  0.127669 -0.005198   
int_mod             0.491205  0.452994  0.563669  0.135180 -0.000829   
wis                 0.473870  0.522324  0.613467  0.201094  0.059840   
wis_mod             0.452507  0.498835  0.589357  0.191398  0.060758   
cha                 0.571148  0.595869  0.671867  0.218002  0.060409   
cha_mod             0.557089  0.569487  0.651243  0.202558  0.058438   
str_save            0.231540  0.336613  0.309616  0.123322  0.135325   
dex_save            0.454821  0.500143  0.540818  0.174371  0.078587   
con_save            0.507516  0.663999  0.666085  0.213263  0.163712   
int_save            0.220386  0.281014  0.395367 -0.039535 -0.009571   
wis_save            0.537626  0.722092  0.791720  0.149805  0.124393   
cha_save            0.477352  0.590412  0.628458  0.143800  0.135087   
history             0.087521  0.149397  0.259376 -0.003762  0.008553   
perception          0.482562  0.579358  0.607297  0.206519  0.104857   
stealth             0.159312  0.069306  0.073048  0.135183  0.024359   
persuasion          0.167344  0.204829  0.216259  0.077149  0.082657   
insight             0.254001  0.299252  0.346519  0.091395  0.093847   
deception           0.159437  0.169126  0.219698  0.051735 -0.036755   
arcana              0.082493  0.180117  0.280647  0.013701 -0.051535   
religion            0.052027  0.099673  0.119612  0.049510 -0.044061   
acrobatics          0.065367 -0.030539 -0.014661  0.091172 -0.050462   
athletics           0.118756  0.051138  0.041790  0.070494 -0.042543   
intimidation        0.228655  0.233970  0.272365  0.068950 -0.065300   
legendary           0.353649  0.489883  0.484703  0.077730  0.136563   

                         fly     climb    burrow  passive_perception  \
ac                  0.334696  0.005731  0.203993            0.551881   
hp                  0.365880 -0.005170  0.212331            0.638400   
cr                  0.354818 -0.026808  0.158885            0.679880   
speed              -0.015583  0.091807  0.130051            0.232996   
swim                0.058982 -0.092373  0.005631            0.132180   
fly                 1.000000 -0.024390  0.154691            0.439707   
climb              -0.024390  1.000000  0.034427            0.065169   
burrow              0.154691  0.034427  1.000000            0.150771   
passive_perception  0.439707  0.065169  0.150771            1.000000   
darkvision          0.221654  0.073030  0.141351            0.259168   
truesight           0.170529 -0.077061 -0.047111            0.333

In [68]:
more_monsters.nunique()

name                  762
size                    6
type                   71
alignment              17
languages             197
ac                     20
hp                    159
cr                     31
speed                  11
swim                   11
fly                    11
climb                   7
burrow                  8
passive_perception     26
darkvision              6
truesight               4
tremorsense             4
blindsight              6
strength               30
str_mod                15
dex                    24
dex_mod                15
con                    24
con_mod                12
intel                  27
int_mod                14
wis                    26
wis_mod                14
cha                    29
cha_mod                15
str_save               17
dex_save               13
con_save               16
int_save               16
wis_save               17
cha_save               18
history                14
perception             20
stealth     

## Visualize

Create any visualizations for your EDA here. Make note in the form of code comments of what your thought process is for your visualizations.

## Summarize Your Results

With your EDA complete, answer the following questions.

1. Was there anything surprising about your dataset? 
2. Do you have any concerns about your dataset? 
3. Is there anything you want to make note of for the next phase of your analysis, which is cleaning data? 

 The **items** dataframe contains a lot of nulls.  not sure if i should clean them by replacing them with something or if i should leave them.
 
 The **more_monsters** dataframe contains monsters from more resources (762 rows compared to **monsters** with 324)
 
 I think I will use **items** and **more_monsters** for the rest of this assignment